cdc/sinkv2/ddlsink/mq/mq_ddl_sink.go

```diff
@@ -0,0 +1,166 @@
+var _ ddlsink.DDLEventSink = (*ddlSink)(nil)
+
+type ddlSink struct {
+	// id indicates which processor (changefeed) this sink belongs to.
+	id model.ChangeFeedID
+	// role indicates what this sink is used for.
+	role util.Role
+	// protocol indicates the protocol used by this sink.
+	protocol config.Protocol
+	// eventRouter used to route events to the right topic and partition.
+	eventRouter *dispatcher.EventRouter
+	// topicManager used to manage topics.
+	// It is also responsible for creating topics.
+	topicManager manager.TopicManager
+	// encoderBuilder builds encoder for the sink.
+	encoderBuilder codec.EncoderBuilder
+	// producer used to send events to the MQ system.
+	// Usually it is a sync producer.
+	producer producer.Producer
+}
+
+func newDDLSink(ctx context.Context,
+	producer producer.Producer,
+	topicManager manager.TopicManager,
+	eventRouter *dispatcher.EventRouter,
+	encoderConfig *codec.Config,
+) (*ddlSink, error) {
+	changefeedID := contextutil.ChangefeedIDFromCtx(ctx)
+	role := contextutil.RoleFromCtx(ctx)
+
+	encoderBuilder, err := codec.NewEventBatchEncoderBuilder(ctx, encoderConfig)
+	if err != nil {
+		return nil, cerror.WrapError(cerror.ErrKafkaInvalidConfig, err)
+	}
+
+	s := &ddlSink{
+		id:             changefeedID,
+		role:           role,
+		protocol:       encoderConfig.Protocol(),
+		eventRouter:    eventRouter,
+		topicManager:   topicManager,
+		encoderBuilder: encoderBuilder,
+		producer:       producer,
+	}
+
+	return s, nil
+}
+
+func (k *ddlSink) WriteDDLEvent(ctx context.Context, ddl *model.DDLEvent) error {
+	encoder := k.encoderBuilder.Build()
+	msg, err := encoder.EncodeDDLEvent(ddl)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if msg == nil {
+		return nil
+	}
+
+	topic := k.eventRouter.GetTopicForDDL(ddl)
+	partitionRule := k.eventRouter.GetDLLDispatchRuleByProtocol(k.protocol)
+	log.Debug("emit ddl event",
+		zap.Uint64("commitTs", ddl.CommitTs),
+		zap.String("query", ddl.Query),
+		zap.String("namespace", k.id.Namespace),
+		zap.String("changefeed", k.id.ID),
+		zap.Any("role", k.role))
+	if partitionRule == dispatcher.PartitionAll {
+		partitionNum, err := k.topicManager.GetPartitionNum(topic)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		err = k.producer.SyncBroadcastMessage(ctx, topic, partitionNum, msg)
+		return errors.Trace(err)
+	}
+	// Notice: We must call GetPartitionNum here,
+	// which will be responsible for automatically creating topics when they don't exist.
+	// If it is not called here and kafka has `auto.create.topics.enable` turned on,
+	// then the auto-created topic will not be created as configured by ticdc.
+	_, err = k.topicManager.GetPartitionNum(topic)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = k.producer.SyncSendMessage(ctx, topic, dispatcher.PartitionZero, msg)
+	return errors.Trace(err)
+}
+
+func (k *ddlSink) WriteCheckpointTs(ctx context.Context,
+	ts uint64, tables []model.TableName,
+) error {
+	encoder := k.encoderBuilder.Build()
```

Comment:
Should we add a rate limiter for writing checkpoint Ts?

Reply:
Maybe, but this no the goal of this PR and it doesn't seem like we've bothered with this kind of operation in the past.

Reply:
Opened an issue https://github.com/pingcap/tiflow/issues/6484

Reply:
We send checkpointTs per second.  So maybe it's OK.